crates/vm/src/lib.rs

```diff
@@ -0,0 +1,197 @@
+//!
+//! ## Execution Future
+//!
+//! The `Vm::exec_*` functions all return `Future`s that not only yield on
+//! async operations, but yield based on a user-specified gas limit too. See the
+//! [`ExecFuture`] docs for further details on the implementation.
+#![deny(missing_docs, unsafe_code)]
+
+pub use access::{mut_keys, mut_keys_set, mut_keys_slices, Access};
+pub use cached::LazyCache;
+#[doc(inline)]
+pub use essential_asm::{self as asm, Op};
+pub use essential_types as types;
+use essential_types::ContentAddress;
+#[doc(inline)]
+pub use future::ExecFuture;
+#[doc(inline)]
+pub use memory::Memory;
+#[doc(inline)]
+pub use op_access::OpAccess;
+#[doc(inline)]
+pub use repeat::Repeat;
+#[doc(inline)]
+pub use stack::Stack;
+#[doc(inline)]
+pub use state_read::StateRead;
+#[doc(inline)]
+pub use total_control_flow::ProgramControlFlow;
+#[doc(inline)]
+pub use vm::Vm;
+
+mod access;
+mod alu;
+pub mod bytecode;
+mod cached;
+mod crypto;
+pub mod error;
+mod future;
+mod memory;
+mod op_access;
+mod pred;
+mod repeat;
+mod sets;
+mod stack;
+mod state_read;
+pub mod sync;
+mod total_control_flow;
+mod vm;
+
+/// Shorthand for the `BytecodeMapped` type representing a mapping to/from [`Op`]s.
+pub type BytecodeMapped<Bytes = Vec<u8>> = bytecode::BytecodeMapped<Op, Bytes>;
+/// Shorthand for the `BytecodeMappedSlice` type for mapping [`Op`]s.
+pub type BytecodeMappedSlice<'a> = bytecode::BytecodeMappedSlice<'a, Op>;
+/// Shorthand for the `BytecodeMappedLazy` type for mapping [`Op`]s.
+pub type BytecodeMappedLazy<I> = bytecode::BytecodeMappedLazy<Op, I>;
+
+/// Unit used to measure gas.
+pub type Gas = u64;
+
+/// Gas limits.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct GasLimit {
+    /// The amount that may be spent synchronously until the execution future should yield.
+    pub per_yield: Gas,
+    /// The total amount of gas that may be spent.
+    pub total: Gas,
+}
+
+/// Distinguish between sync and async ops to ease `Future` implementation.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
+pub(crate) enum OpKind {
+    /// Operations that yield immediately.
+    Sync(OpSync),
+    /// Operations returning a future.
+    Async(OpAsync),
+}
+
+/// The set of operations that are performed asynchronously.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
+pub struct OpAsync(asm::StateRead);
+
+/// The set of operations that are performed synchronously.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
+pub enum OpSync {
+    /// `[asm::Access]` operations.
+    Access(asm::Access),
+    /// `[asm::Alu]` operations.
+    Alu(asm::Alu),
+    /// `[asm::TotalControlFlow]` operations.
+    ControlFlow(asm::TotalControlFlow),
+    /// `[asm::Crypto]` operations.
+    Crypto(asm::Crypto),
+    /// `[asm::Memory]` operations.
+    Memory(asm::Memory),
+    /// `[asm::Pred]` operations.
+    Pred(asm::Pred),
+    /// `[asm::Stack]` operations.
+    Stack(asm::Stack),
+}
```

Comment:
Would it be useful to mention in the high level docs, or anywhere else you see fit, that the all ops = async (state read) ops + sync (all else) ops?

Reply:
Oh yeah, I meant in this VM implementation. Passing this is fine though!